src/ert/_c_wrappers/enkf/model_config.py

```diff
@@ -1,299 +1,122 @@
-        return self._get_time_map()
+    def get_last_history_restart(self) -> int:
+        if self.refcase:
+            return self.refcase.last_report
+        if self.time_map:
+            return self.time_map.last_step()
+        return -1
 
     def __repr__(self):
-        return f"<ModelConfig({self.__str__()})>"
+        return f"ModelConfig(\n{self.__str__()}\n)"
 
     def __str__(self):
         return (
-            (
-                f"data_root: {self.data_root()},\n"
-                f"num_realizations: {self.num_realizations},\n"
-                f"obs_config_file: {self.obs_config_file},\n"
-                f"ens_path: {self.getEnspath()},\n"
-                f"run_path_format: {self.getRunpathFormat()},\n"
-                f"job_name_format: {self.getJobnameFormat()},\n"
-                f"run_path: {self.getRunpathAsString()},\n"
-                f"max_internal_submit: {self.get_max_internal_submit()},\n"
-                f"gen_kw_export_name: {self.getGenKWExportName()},\n"
-                f"time_map: {self.get_time_map()},\n"
-            )
-            if self._address()
-            else ""
+            f"num_realizations: {self.num_realizations},\n"
+            f"refcase: {self.refcase},\n"
+            f"config_dir: {self._config_dir},\n"
+            f"data_root: {self.data_root},\n"
+            f"ens_path: {self.ens_path},\n"
+            f"history_source: {self.history_source},\n"
+            f"runpath_format_string: {self.runpath_format_string},\n"
+            f"job_name_format: {self.jobname_format_string},\n"
+            f"gen_kw_export_name: {self.gen_kw_export_name},\n"
+            f"obs_config_file: {self.obs_config_file},\n"
+            f"time_map_file: {self._time_map_file}"
         )
 
-    def __ne__(self, other):
-        return not self == other
-
     def __eq__(self, other):
-        if os.path.normpath(self.data_root()) != os.path.normpath(other.data_root()):
-            return False
-
-        if self.num_realizations != other.num_realizations:
-            return False
-
-        if os.path.realpath(self.obs_config_file) != os.path.realpath(
-            other.obs_config_file
-        ):
-            return False
-
-        if os.path.realpath(self.getEnspath()) != os.path.realpath(other.getEnspath()):
-            return False
+        checks = [
+            self.num_realizations == other.num_realizations,
+            self._config_dir == other._config_dir,
+            self.data_root == other.data_root,
+            self.ens_path == other.ens_path,
+            self.history_source == other.history_source,
+            self.runpath_format_string == other.runpath_format_string,
+            self.jobname_format_string == other.jobname_format_string,
+            self.gen_kw_export_name == other.gen_kw_export_name,
+            self.obs_config_file == other.obs_config_file,
+            self._time_map_file == other._time_map_file,
+            self.time_map == other.time_map,
+        ]
```

Comment:
Very lovely:+1:, but this makes debugging when there is a mismatch a bit harder to identify where it happened. I like it it looks a bit cleaner than all the `ifs`.

Reply:
i know what you mean - i resort to commenting out blocks of the checks at a time to figure out where the error is. does that sound like an OK alternative? one can do a binary search that way, and arrive at the right one in O(log(number of checks)) time [finally a chance to use big o notation in real life] ; )

Reply:
>i resort to commenting out blocks of the checks at a time to figure out where the error is. does that sound like an OK alternative?

That is fine